app/models/manageiq/providers/openstack/identity_sync_mixin.rb

```diff
@@ -0,0 +1,183 @@
+module ManageIQ::Providers::Openstack::IdentitySyncMixin
+  def list_users
+    # V2 requires adminURL as endpoint_type
+    connection_options = {:service => "Identity", :openstack_endpoint_type => 'adminURL'}
+    ext_management_system.with_provider_connection(connection_options) do |service|
+      service.list_users.body["users"]
+    end
+  end
+
+  def new_users
+    users = []
+    openstack_users = list_users
+    openstack_users.each do |u|
+      username = u["name"]
+      next if skip_user?(username)
+      user = User.find_by(:userid => username)
+      users << u if user.nil?
+    end
+    users
+  end
+
+  def sync_users_queue(userid, admin_role_id, member_role_id, password_digest)
+    task_opts = {
+      :action => "Sync Users",
+      :userid => userid
+    }
+    queue_opts = {
+      :class_name  => self.class.name,
+      :method_name => 'sync_users',
+      :instance_id => id,
+      :role        => 'ems_operations',
+      :zone        => ext_management_system.my_zone,
+      :args        => [admin_role_id, member_role_id, password_digest]
+    }
+    MiqTask.generic_action_with_callback(task_opts, queue_opts)
+  end
+
+  def sync_users(admin_role_id, member_role_id, password_digest)
+    myusers = list_users
+    myusers.each do |u|
+      email = u["email"]
+      username = u["name"]
+      user_uuid = u["id"]
+      enabled = u["enabled"]
+
+      next if skip_user?(username) || enabled == false
+
+      user = create_or_find_user(user_uuid, username, email, password_digest)
+      # user is nil if an exist user exist with the same username but different email
+      # in this case we don't do anything
+      next if user.nil?
+
+      user_projects = keystone.list_user_projects_tenants(user_uuid)
+      user_projects.each do |p|
+        project_name = p["name"]
+        project_uuid = p["id"]
+
+        # skip service tenant
+        next if project_name == "service"
+
+        cloud_tenant = CloudTenant.find_by(:name => project_name, :ems_id => id)
+        next if cloud_tenant.nil?
+        tenant = Tenant.find_by(:source_id => cloud_tenant.id, :source_type => 'CloudTenant')
+        next if tenant.nil?
+
+        # Find roles that this user has for this project/tenant
+        roles = keystone.list_project_tenant_user_roles(project_uuid, user_uuid)
+        roles.each do |r|
+          role_name = r["name"]
+          create_or_find_miq_group_and_add_user(user, tenant, role_name, admin_role_id, member_role_id)
+        end
+      end
+    end
+  end
+
+  def keystone
+    openstack_handle.identity_service
+  end
+
+  def skip_user?(username)
+    users_to_skip = ["admin", "neutron", "heat", "aodh", "cinder", "swift", "glance", "placement", "gnocchi", "nova", "heat-cfn", "panko", "ceilometer"]
```

Comment:
I tested against an undercloud-used-as-a-cloud-provider, and noticed the following users were also created: mistral, zaqar-websocket, ironic, ironic-inspector, zaqar. I know that there are some plans in bringing these services into the overcloud - would it be possible to add them to this list?